sql, repr: Add bytes type for sql value

moulimukherjee opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->
For COPY to S3 SQL ([design doc](https://github.com/moulimukherjee/materialize/blob/design-doc-copy-to-s3/doc/developer/design/20231117_copy_to_s3.md#option-max-file-size)) we'll need a new option type to accept bytes either as a formatted string like '120 MB' or an integer for the number of bytes. 

This PR adds such a type and is also updating existing `max_query_result_size` to use this type.

From postgres' config settings https://www.postgresql.org/docs/current/config-setting.html#CONFIG-SETTING-NAMES-VALUES

> Numeric with Unit: Some numeric parameters have an implicit unit, because they describe quantities of memory or time. The unit might be bytes, kilobytes, blocks (typically eight kilobytes), milliseconds, seconds, or minutes. An unadorned numeric value for one of these settings will use the setting's default unit, which can be learned from pg_settings.unit. For convenience, settings can be given with a unit specified explicitly, for example '120 ms' for a time value, and they will be converted to whatever the parameter's actual unit is. Note that the value must be written as a string (with quotes) to use this feature. The unit name is case-sensitive, and there can be whitespace between the numeric value and the unit.

> Valid memory units are B (bytes), kB (kilobytes), MB (megabytes), GB (gigabytes), and TB (terabytes). The multiplier for memory units is 1024, not 1000.

> Valid time units are us (microseconds), ms (milliseconds), s (seconds), min (minutes), h (hours), and d (days).

> If a fractional value is specified with a unit, it will be rounded to a multiple of the next smaller unit if there is one. For example, 30.1 GB will be converted to 30822 MB not 32319628902 B. If the parameter is of integer type, a final rounding to integer occurs after any unit conversion.

### Motivation
Fixes https://github.com/MaterializeInc/materialize/issues/24217

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer
Comments inline

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

benesch commented on src/sql-parser/src/ast/defs/value.rs:
```diff
@@ -238,3 +239,211 @@ impl Default for IntervalValue {
 }
+
+pub enum BytesUnit {
+    B,
+    KB,
+    MB,
+    GB,
+    TB,
+    PB,
```
Double check me on this, but I believe Rust capitalization rules for acronyms yield "Kb", "Gb" etc.

```suggestion
    B,
    Kb,
    Mb,
    Gb,
    Tb,
    Pb,
```

moulimukherjee replied:
Yes, Rust does camel case for acronyms. In this case I was not sure since `Gb` actually means something different than `GB`

benesch commented on src/sql-parser/src/ast/defs/value.rs:
```diff
@@ -238,3 +239,211 @@ impl Default for IntervalValue {
+        assert_eq!(parse("8 PB"), BytesValue::pb(8).as_u64());
+        assert_eq!(parse("81 PB"), BytesValue::pb(81).as_u64());
+    }
+}
```
Since the SQL parser won't ever parse things as a `BytesValue` directly—it'll always keep them as strings—better to put these types somewhere else! I think `mz_repr` is probably the right place.

moulimukherjee replied:
Moved under `mz_repr`. Also renamed it to `Bytes` (I had done `BytesValue` to follow the pattern for `IntervalValue`).

benesch commented on test/testdrive/session.td:
```diff
@@ -49,9 +49,9 @@ max_kafka_connections               1000                    "The maximum number
 max_objects_per_schema              1000                    "The maximum number of objects in a schema (Materialize)."
 max_postgres_connections            1000                    "The maximum number of PostgreSQL connections in the region, across all schemas (Materialize)."
-max_query_result_size               1073741824              "The maximum size in bytes for a single query's result (Materialize)."
+max_query_result_size               "1.0 GB"              "The maximum size in bytes for a single query's result (Materialize)."
```
PostgreSQL doesn't put a space between the value and the unit:

```
benesch=# set log_temp_files = '10 GB';
SET
benesch=# show log_temp_files;
 log_temp_files 
----------------
 10GB
(1 row)
```

moulimukherjee replied:
updated

benesch replied:
(Looks to still have the space!)

moulimukherjee replied:
🤦‍♀️

moulimukherjee replied:
Updated for real this time!

moulimukherjee commented on src/repr/src/bytes.rs:
```diff
@@ -0,0 +1,238 @@
+/// Valid units are B, KB, MB, GB, TB, PB with multiples of 1024
+/// where 1MB = 1024 KB
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize, Default)]
+pub struct Bytes(pub u64);
```
This name seemed more appropriate, happy to take other suggestions too.

ParkMyCar replied:
Personally I would prefer `ByteSize` or `ByteAmount` since [`Bytes`](https://crates.io/crates/bytes) is a commonly used struct that is a container for blobs of bytes

benesch commented on src/repr/src/bytes.rs:
```diff
@@ -0,0 +1,243 @@
+/// [`bytesize::ByteSize`]. Instead of MiB or GiB and so on, it uses MB, GB for the sizes
+/// with 1024 multiplier.
+/// Valid units are B, KB, MB, GB, TB, PB with multiples of 1024
+/// where 1MB = 1024 KB
```
Thanks for exactly matching PostgreSQL semantics here.

benesch commented on src/repr/src/bytes.rs:
```diff
@@ -0,0 +1,243 @@
+//
+// This file is derived from the sqlparser-rs project, available at
+// https://github.com/andygrove/sqlparser-rs. It was incorporated
+// directly into Materialize on December 21, 2019.
```
This doesn't look like the right copyright header.

moulimukherjee replied:
fixed 🤦‍♀️

benesch commented on src/repr/src/bytes.rs:
```diff
@@ -0,0 +1,243 @@
+    pub fn to_string(bytes: u64) -> String {
+        let unit = 1024;
+        let unit_base = 6.907755279; // ln 1000
+        let unit_prefix = ['K', 'M', 'G', 'T', 'P', 'E'];
```
PostgreSQL seems to use `kB`, not `KB`. Also `E` isn't a thing that PostgreSQL seems to support?

moulimukherjee replied:
The 'E' was copy pasta from ByteSize, removed it. Updated to `kB` everywhere.

benesch commented on src/repr/src/bytes.rs:
```diff
@@ -0,0 +1,243 @@
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s.to_uppercase().as_ref() {
+            "B" => Ok(Self::B),
+            "KB" => Ok(Self::Kb),
```
```suggestion
            "kB" => Ok(Self::Kb),
```

moulimukherjee replied:
Here it's comparing after `to_uppercase`, but fixed the to_string at other places.

ParkMyCar commented on src/repr/src/bytes.rs:
```diff
@@ -0,0 +1,232 @@
+/// with 1024 multiplier.
+/// Valid units are B, kB, MB, GB, TB, PB with multiples of 1024
+/// where 1MB = 1024kB
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize, Default)]
```
deriving `proptest_derive::Arbitrary` here would be great! Then we can very easily get some randomized testing coverage

ParkMyCar commented on src/repr/src/bytes.rs:
```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+impl FromStr for Bytes {
```
Should `Bytes` roundtrip through it's string representation? If so can we add a `proptest` for that? Something like
```
proptest! {
  #[mz_ore::test]
  fn proptest_bytes_roundtrips_string(og: Bytes) {
    let s = og.to_string();
    let rnd = Bytes::from_str(&s).expect("roundtrip");
    prop_assert_eq!(og, rnd);
  }
}
```

moulimukherjee replied:
Yup, it should roundtrip! Let me add that

moulimukherjee replied:
I spoke too soon. It actually does not with fractional values. For ex `1025 bytes`, will still be shown as `1 kB`, which roundtrips back to `1024 bytes`. There are some lossy casts `u64` to `f64` and vice versa

moulimukherjee replied:
I added a String -> ByteSize - String roundtrip instead.

ParkMyCar replied:
Makes sense, thanks!

ParkMyCar reviewed (approved):
Thanks for renaming `Bytes` to `ByteSize`!

ParkMyCar reviewed (changes requested):
@moulimukherjee I chatted with @benesch and let’s make sure we’re totally matching Postgres semantics here, especially with the to_string conversion. I’m on mobile right now, but happy to help dig up the specifics!

moulimukherjee commented:
@ParkMyCar I think I matched everything with their documented spec but let me run some sql and compare with postgres.

moulimukherjee commented:
Will also validate 
> If a fractional value is specified with a unit, it will be rounded to a multiple of the next smaller unit if there is one. For example, 30.1 GB will be converted to 30822 MB not 32319628902 B. If the parameter is of integer type, a final rounding to integer occurs after any unit conversion.

moulimukherjee commented:
So, I ran some comparisons. For a setting, Postgres has both a type (float, integer etc.) and an unit (kB, MB etc.). 

For example, for `log_temp_files`, the type is `integer` and unit is `kB`. So even if say `500 B` is specified, it's getting rounded in `kB`. Also, since the type is `integer`, it's always going to show an integer even if a fraction is given using the rounding mentioned above (i.e. 30.1 GB will be converted to 30822 MB).

I am assuming we don't want to introduce fully fledged 'type' and 'unit' for each setting. Rather for us, we can have this assumption, that for a size with bytes `type` is always `integer` and our unit is always `B`. This would mean that something like `30.9 B` will be rounded to `31B`, but `30.9 kB` will be rounded to `31642B`.

@benesch @ParkMyCar Does the above sound okay to you? I will update the implementation then.

moulimukherjee commented on src/repr/src/bytes.rs:
```diff
@@ -0,0 +1,327 @@
+    Arbitrary, Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize, Default,
+)]
+pub struct ByteSize {
+    size: u32,
```
I changed this to `u32` from `u64`, because while parsing since the number can have fractions, I am using `f64` which can hold less number of bits than `u64`. But this also means we are limiting the size specified to `u32::MAX` (It does not seem that bad though, people can always leverage higher units to represent bigger numbers).

The other option might be to keep this `u64` here and use some Decimal library while parsing from string.

moulimukherjee replied:
Not applicable anymore, refactored to store the entire bytes as `u64`

moulimukherjee commented:
I have changed the implementation completely and added more tests. This should now do rounding correctly. And also, removed `PB`, since postgres supports till `TB`.

moulimukherjee commented:
Total aside, I just noticed that the way we parse `Duration` https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/session/vars.rs#L4837, it does not do the rounding correctly.
For eg, postgres will round it down to the lower unit similar to bytes
```
postgres=# set log_min_duration_sample = '10.2min';
SET
postgres=# show log_min_duration_sample;
 log_min_duration_sample
-------------------------
 612s
```
And some minor differences like there shouldn't be any space between the number and the unit. Is this something which should be fixed as well? (If so, I can take that as well) Or it's risky since it's already out there?

benesch commented:
> So, I ran some comparisons. For a setting, Postgres has both a type (float, integer etc.) and an unit (kB, MB etc.).
> 
> For example, for `log_temp_files` (https://postgresqlco.nf/doc/en/param/log_temp_files/), the type is `integer` and unit is `kB`. So even if say `500 B` is specified, it's getting rounded in `kB`. Also, since the type is `integer`, it's always going to show an integer even if a fraction is given using the rounding mentioned above (i.e. 30.1 GB will be converted to 30822 MB).
> 
> I am assuming we don't want to introduce fully fledged 'type' and 'unit' for each setting. Rather for us, we can have this assumption, that for a size with bytes `type` is always `integer` and our base unit is always `B`. This would mean that something like `30.9 B` will be rounded to `31B`, but `30.9 MB` will be rounded to `31642kB`.
> 
> @benesch @ParkMyCar Does the above sound okay to you?

Yes, I think so! I think one day we'll want to support exactly matching PostgreSQL's behavior of associating a type/unit to each bytes-valued parameter. But as long as all of our current bytes-valued parameters are custom to Materialize, I think we're fine to do the simpler thing for now. As far as I can tell the current bytes-valued parameters are

  * `max_copy_from_size`
  * `max_query_result_size`
  * `max_result_size`

which are all custom Materialize parameters. 

> Rather for us, we can have this assumption, that for a size with bytes `type` is always `integer` and our base unit is always `B`. This would mean that something like `30.9 B` will be rounded to `31B`, but `30.9 MB` will be rounded to `31642kB`.

👍🏽

nrainer-materialize commented:
I triggered a [nightly](https://buildkite.com/materialize/nightlies/builds/5997) and a [coverage](https://buildkite.com/materialize/coverage/builds/342) build; I will report the results when they are ready.

nrainer-materialize commented:
> I triggered a [nightly](https://buildkite.com/materialize/nightlies/builds/5997) and a [coverage](https://buildkite.com/materialize/coverage/builds/342) build; I will report the results when they are ready.

There is one failure in Nightly that I would like you to have a look at: https://buildkite.com/materialize/nightlies/builds/5997#018d1183-6f1c-4537-a3a6-30efe36f2bb2
Other failures are expected if you are not on the latest state of main.

Coverage looks good besides one gap that I will annotate. 👍

nrainer-materialize commented on src/sql/src/plan/with_options.rs:
```diff
@@ -180,6 +181,26 @@ impl ImpliedValue for Duration {
 }
 
+impl TryFromValue<Value> for ByteSize {
+    fn try_from_value(v: Value) -> Result<Self, PlanError> {
+        match v {
+            Value::Number(value) | Value::String(value) => Ok(value
+                .parse::<ByteSize>()
+                .map_err(|e| sql_err!("invalid bytes value: {e}"))?),
+            _ => sql_bail!("cannot use value as bytes"),
+        }
+    }
```
Not covered, can we add a test?

moulimukherjee replied:
Yeah, this is expected to be not covered. We don't have a use of this as a sql option yet. This will be used for the COPY to S3 sql. I can remove this, and make it part of the copy to s3 PR instead where it can be tested.

nrainer-materialize replied:
No, it's okay!

moulimukherjee replied:
Since the other PR got merged first, I added a usage of this as a sql option. This part should now be covered by mzcompose test.

moulimukherjee commented on test/launchdarkly/mzcompose.py:
```diff
@@ -84,7 +84,7 @@ def workflow_default(c: Composition) -> None:
         # Assert that the default max_result_size is served when sync is disabled.
         with c.override(Materialized()):
             c.up("materialized")
-            c.testdrive("\n".join(["> SHOW max_result_size", "1073741824"]))
```
@nrainer-materialize I updated the failing launchdarkly test. This was failing since now the value is displayed with a unit.

nrainer-materialize replied:
Thank you!

ParkMyCar reviewed (approved):
Thanks for iterating on this Mouli! I'm happy with where we landed :)
